File: 程式設計：資料結構答案檔/Python/303.py
def generate_combinations(items):  
    combinations = []
    for i in range(len(items)):
        for j in range(i + 1, len(items)):
            combinations.append(sorted([items[i],  items[j]]))
    return combinations

def compute(order, relations, combinations):
    for i in range(len(relations)):
        if  combinations[i][0] in order and combinations[i][1] in order:
            relations[i] += 1

def sortRelation(relations, combinations):
    for i in range(len(relations)-1):
        for j in range(len(relations)-i-1):
            if relations[j]<relations[j+1] or (relations[j]==relations[j+1] and combinations[j]<combinations[j+1]) : 
                relations[j], relations[j+1] = relations[j+1], relations[j]
                combinations[j], combinations[j+1] = combinations[j+1], combinations[j]
    for i in range(3):
        for c in combinations[i]:
            print(c, end=' ')       
        print(relations[i])       
     
def check():
    items =['apple', 'watermelon', 'pawpaw', 'banana', 'pineapple']
    combinations = generate_combinations(items)
    relations = [0 for i in range(len(combinations))]
    n = int(input())
    for i in range(n):
        order = input().split()
        compute(order, relations,combinations)
    
    sortRelation(relations, combinations)
    
if __name__ == "__main__":
    check()

File: 程式設計：資料結構答案檔/C++/303.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <string>

using namespace std;

vector<pair<string, string>> generateCombinations(const vector<string>& items) {
    vector<pair<string, string>> combinations;
    for (size_t i = 0; i < items.size(); ++i) {
        for (size_t j = i + 1; j < items.size(); ++j) {
            if (items[i] < items[j])
                combinations.push_back(make_pair(items[i], items[j]));
            else
                combinations.push_back(make_pair(items[j], items[i]));
        }
    }
    return combinations;
}

void compute(const set<string>& order, vector<int>& relations, const vector<pair<string, string>>& combinations) {
    for (size_t i = 0; i < combinations.size(); ++i) {
        if (order.find(combinations[i].first) != order.end() && order.find(combinations[i].second) != order.end()) {
            relations[i]++;
        }
    }
}

bool sortRelation(const pair<pair<string, string>, int>& a, const pair<pair<string, string>, int>& b) {
    if (a.second == b.second)
        return a.first > b.first;
    return a.second > b.second;
}

int main() {
    vector<string> items = {"apple", "watermelon", "pawpaw", "banana", "pineapple"};
    auto combinations = generateCombinations(items);
    vector<int> relations(combinations.size(), 0);
    
    int n;
    cin >> n;
    cin.ignore();

    for (int i = 0; i < n; ++i) {
        string line;
        getline(cin, line);
        set<string> order;
        size_t pos = 0;
        while ((pos = line.find(' ')) != string::npos) {
            order.insert(line.substr(0, pos));
            line.erase(0, pos + 1);
        }
        order.insert(line);

        compute(order, relations, combinations);
    }

    vector<pair<pair<string, string>, int>> combined;
    for (size_t i = 0; i < combinations.size(); ++i) {
        combined.push_back(make_pair(combinations[i], relations[i]));
    }

    sort(combined.begin(), combined.end(), sortRelation);

    for (int i = 0; i < 3 && i < combined.size(); ++i) {
        cout << combined[i].first.first << " " << combined[i].first.second << " " << combined[i].second << endl;
    }

    return 0;
}